File: src/api/github.rs
use anyhow::{Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};

pub struct GitHubClient {
    client: Client,
    owner: String,
    repo: String,
    token: String,
}

#[derive(Debug, Serialize)]
struct CreatePullRequestPayload {
    title: String,
    body: String,
    head: String,
    base: String,
}

#[derive(Debug, Deserialize)]
struct PullRequest {
    html_url: String,
    #[allow(dead_code)]
    number: u64,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct Repository {
    full_name: String,
}

impl GitHubClient {
    pub fn new(owner: String, repo: String, token: String) -> Self {
        Self {
            client: Client::new(),
            owner,
            repo,
            token,
        }
    }

    pub async fn create_pull_request(
        &self,
        source_branch: &str,
        target_branch: &str,
        title: &str,
        description: &str,
    ) -> Result<String> {
        let payload = CreatePullRequestPayload {
            title: title.to_string(),
            body: description.to_string(),
            head: source_branch.to_string(),
            base: target_branch.to_string(),
        };

        let url = format!(
            "https://api.github.com/repos/{}/{}/pulls",
            self.owner, self.repo
        );

        let response = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.token))
            .header("Accept", "application/vnd.github+json")
            .header("User-Agent", "devflow-cli")
            .json(&payload)
            .send()
            .await
            .context("Failed to send pull request creation request")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            anyhow::bail!("GitHub API error ({}): {}", status, text);
        }

        let pr = response
            .json::<PullRequest>()
            .await
            .context("Failed to parse pull request response")?;

        Ok(pr.html_url)
    }

    pub async fn get_repo_info(&self) -> Result<String> {
        let url = format!(
            "https://api.github.com/repos/{}/{}",
            self.owner, self.repo
        );

        let response = self
            .client
            .get(&url)
            .header("Authorization", format!("Bearer {}", self.token))
            .header("Accept", "application/vnd.github+json")
            .header("User-Agent", "devflow-cli")
            .send()
            .await
            .context("Failed to fetch repository information")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            anyhow::bail!("GitHub API error ({}): {}", status, text);
        }

        let repo = response
            .json::<Repository>()
            .await
            .context("Failed to parse repository response")?;

        Ok(repo.full_name)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_github_client_creation() {
        let client = GitHubClient::new(
            "owner".to_string(),
            "repo".to_string(),
            "test-token".to_string(),
        );
        assert_eq!(client.owner, "owner");
        assert_eq!(client.repo, "repo");
        assert_eq!(client.token, "test-token");
    }
}

File: internal/providers/github/github.go
package github

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

type Client struct {
	owner   string
	repo    string
	token   string
	http    *http.Client
	baseURL string
}

func NewClient(owner, repo, token string) *Client {
	return &Client{
		owner:   owner,
		repo:    repo,
		token:   token,
		http:    &http.Client{Timeout: 30 * time.Second},
		baseURL: "https://api.github.com",
	}
}

func (c *Client) CreatePullRequest(sourceBranch, targetBranch, title, description string) (string, error) {
	payload := map[string]string{
		"title": title,
		"body":  description,
		"head":  sourceBranch,
		"base":  targetBranch,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}

	url := fmt.Sprintf("%s/repos/%s/%s/pulls", c.baseURL, c.owner, c.repo)
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.token))
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "devflow-cli")

	resp, err := c.http.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("github api error (%d): %s", resp.StatusCode, string(data))
	}

	var result struct {
		HTMLURL string `json:"html_url"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return "", err
	}
	return result.HTMLURL, nil
}
